localstack/services/kms/provider.py

```diff
@@ -344,56 +416,33 @@ def list_grants(
 
         return ListGrantsResponse(Grants=page, **kwargs)
 
-    # Honestly, this is a mess in AWS KMS. Hashtag "do we follow specifications that are a pain to customers or do we
-    # diverge from AWS and make the life of our customers easier?"
-    #
-    # Both RetireGrant and RevokeGrant operations delete a grant. The differences between them are described here:
-    # https://docs.aws.amazon.com/kms/latest/developerguide/grant-manage.html#grant-delete
-    # Essentially:
-    # - Permissions to RevokeGrant are controlled through IAM policies or through key policies, while permissions to
-    # RetireGrant are controlled by settings inside the grant itself.
-    # - A grant to be retired can be specified by its GrantToken or its GrantId/KeyId pair. While revoking grants can
-    # only be done with a GrantId/KeyId pair.
-    # - For RevokeGrant, KeyId can be either an actual key ID, or an ARN of that key. While for RetireGrant only key
-    # ARN is accepted as a KeyId.
-    #
-    # We currently do not model permissions for retirement and revocation of grants. At least not in KMS,
-    # maybe IAM in LocalStack has some modelling though. We also accept both key IDs and key ARNs for both
-    # operations. So apart from RevokeGrant not accepting GrantToken parameter, we treat these two operations the same.
     @staticmethod
-    def _delete_grant(
-        store: KmsStore, grant_id: str = None, key_id: str = None, grant_token: str = None
-    ):
-        if grant_token:
-            if grant_token not in store.grant_tokens:
-                raise NotFoundException(f"Unable to find grant token {grant_token}")
-            grant_id = store.grant_tokens[grant_token]
-            # Do not really care about the key ID if a grant is identified by a token. But since a key has to be
-            # validated when a grant is identified by GrantId/KeyId pair, and since we want to use the same code in
-            # both cases - when we have a grant token or a GrantId/KeyId pair - have to set key_id.
-            key_id = store.grants[grant_id].metadata["KeyId"]
-
-        # KeyId can potentially hold one of multiple different types of key identifiers. Here we find a key no
-        # matter which type of id is used.
-        key = store.get_key(key_id, any_key_state_allowed=True)
-        key_id = key.metadata.get("KeyId")
+    def _delete_grant(store: KmsStore, grant_id: str, key_id: str):
+        grant = store.grants[grant_id]
 
-        if grant_id not in store.grants:
-            raise InvalidGrantIdException()
-        if store.grants[grant_id].metadata["KeyId"] != key_id:
+        _, _, grant_key_id = parse_key_arn(grant.metadata.get("KeyId"))
+        if key_id != grant_key_id:
             raise ValidationError(f"Invalid KeyId={key_id} specified for grant {grant_id}")
 
```

Comment:
This is refactored to simplify the control flow. Only the common logic is retained in this method.

While `RevokeGrant` and `RetireGrant` both delete a grant, their parameters they take are different.